### DISCUSSION TITLE
Idea for emp-tune-dts-plugin

### DISCUSSION THREAD

**Discussion by kevinschaffter** [empjs/emp#79] [Ideas]
Thank you for making `emp-tune-dts-plugin` so far I think it is the best solution for combining module federation + TypeScript.

I am running into an issue where even though I can import types from a remote, not all of those modules are in the webpack module federation `shared` property.

For example my federated remote has:

```ts
export { Button } from './components/button';
export { Text } from './components/text';
```

```ts
exposes: [
./button: ./components/button
]
```

In the host after emp sync there are no TS errors but the app crashes because its not "exposed":
```
import { Text } from 'remote/text'
```

Is there a way to modify your plugin to only create `d.ts` for modules that are actually shared?

**Comment by flyyuan** [comment]
1. Importing all module types first will not cause a crash.
2. To share this project module use "exposes", and to share third-party modules use "share". Because you use share to cause a crash when you "share" the modules of this project, you should use "exposes" to solve the crash.
3. You can refer to the demo to declare the shared module: https://github.com/efoxTeam/emp/blob/main/projects/demo1/emp-config.js

**Reply by kevinschaffter** [reply]
Sorry... I meant exposes.  I changed it above.  (that was a typo)

My problem is tune-dts-plugin creates d.ts for _everything_ including components that are not exposed.... so its hard to know in the host app what is exposed or not because declarations are created for everything in remote.  I was wondering if there is a solution to have tune-dts _only_ create declarations for modules that are in the shared array.

**Reply by flyyuan** [reply]
Can not be excluded temporarily, but does not affect the use. Code with additional type declarations is normal. Only exposes may be generated later.
